### ISSUE TITLE
deepMerge and deepClone

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/dataflower#67]
```js
            deepCopy({
                x: function (subject, source){
                    return deepCopy({y: shallowCopy}, subject, source);
                }
            }, subject, source1, source2, ...)
```
or

```js
            deepCopy({x: {y: shallowCopy}}, subject, source1, source2, ...)
```

**Comment by inf3rno** [comment]
How to use this with merge???

```js
var merge = function (subject, source) {
    if (!(subject instanceof Object))
        throw new InvalidArguments();
    if (subject.merge instanceof Function)
        return subject.merge.apply(subject, Array.prototype.slice.call(arguments, 1));
    return shallowCopy.apply(null, arguments);
};
```

**Comment by inf3rno** [comment]
I think it is already implemented: 

```js
    if (subject.merge instanceof Function)
        return subject.merge.apply(subject, Array.prototype.slice.call(arguments, 1));
```

So this is not a problem...

**Comment by inf3rno** [comment]
It would be nice to have something similar by nested inheritance. The current solution is clone:

```js
var clone = function (subject) {
    if (typeof (subject) != "object" || subject === null)
        return subject;
    if (subject instanceof Array)
        return subject.slice();
    if (subject instanceof Date)
        return new Date(subject);
    if (subject instanceof RegExp)
        return new RegExp(subject);
    if (subject.clone instanceof Function)
        return subject.clone();
    return Object.create(subject);
};
```

**Comment by inf3rno** [comment]
```js
var shallowClone = function (subject) {
    if (typeof (subject) != "object" || subject === null)
        return subject;
    if (subject instanceof Array)
        return subject.slice();
    if (subject instanceof Date)
        return new Date(subject);
    if (subject instanceof RegExp)
        return new RegExp(subject);
    if (subject.clone instanceof Function)
        return subject.clone();
    return Object.create(subject);
};
```

**Comment by inf3rno** [comment]
Subscription is a great example:

```js
    merge: function (source) {
        for (var sourceIndex in arguments) {
            source = arguments[sourceIndex];
            if (source === undefined || source === null)
                continue;
            if (!(source instanceof Object))
                throw new InvalidArguments();

            if (source.items !== undefined) {
                if (!(source.items instanceof Array))
                    throw new Subscription.ItemsRequired();
                if (source.items.length)
                    this.addAll.apply(this, source.items);
            }

            var backup = {
                items: this.items
            };

            shallowCopy(this, source);
            shallowCopy(this, backup);
        }
        return this;
    },
```

Another example is Wrapper:

```js
    build: function () {
        Base.prototype.build.call(this);
        this.preprocessors = clone(this.preprocessors);
        this.properties = clone(this.properties);
    },
    merge: function (source) {
        for (var sourceIndex in arguments) {
            source = arguments[sourceIndex];
            if (source === undefined || source === null)
                continue;
            if (!(source instanceof Object))
                throw new InvalidArguments();

            if (source.preprocessors !== undefined) {
                if (!(source.preprocessors instanceof Array))
                    throw new Wrapper.ArrayRequired();
                for (var preprocessorIndex in source.preprocessors)
                    if (!(source.preprocessors[preprocessorIndex] instanceof Function))
                        throw new Wrapper.PreprocessorRequired();
                this.preprocessors.push.apply(this.preprocessors, source.preprocessors);
            }

            if (source.done !== undefined && !(source.done instanceof Function))
                throw new Wrapper.FunctionRequired();

            if (source.algorithm !== undefined && !(source.algorithm instanceof Function))
                throw new Wrapper.AlgorithmRequired();

            if (source.properties !== undefined) {
                if (!(source.properties instanceof Object))
                    throw new Wrapper.PropertiesRequired();
                shallowCopy(this.properties, source.properties);
            }

            var backup = {
                preprocessors: this.preprocessors,
                properties: this.properties
            };
            shallowCopy(this, source);
            shallowCopy(this, backup);
        }
        return this;
    },
```

I have to solve these issues.

**Comment by inf3rno** [comment]
I have to rethink merge, shallowCopy, clone . I'll add deepCopy, deepClone, so I'll probably need a shallowClone as well and clone will be something like merge. So it won't die in an infinite loop, at least if I remember well that's why I defined merge instead of using a single shallowCopy... I think this should be done before #66 so it will be easier to rethink that part with existing examples.

**Comment by inf3rno** [comment]
I think deepCopy should work the same way as shallowCopy by default.
If it does not get global as context, then it uses the instructions to deep copy the subject.

```js
deepCopy({a:{b:1}}, {a:{c:2}})
// {a:{c:2}}
```

```js
deepCopy.call({a:deepCopy},{a:{b:1}}, {a:{c:2}})
// {a:{b:1,c:2}}
```

```js
deepCopy.call({"*": deepCopy},{a:{b:1}}, {a:{c:2}})
// {a:{b:1,c:2}}
```

```js
deepCopy.call(function (subject, source){
    if (subject instanceof Object) 
        return deepCopy(subject, source) 
    else 
        return source;
},{a:{b:1}}, {a:{c:2}})
```

**Comment by inf3rno** [comment]
```js
deepCopy.call(function (a,b){return a+b;}, 1, 2, 3);
// 6
```
Should I use this?
or
```js
deepCopy.call(function (a,b,c){return a+b+c;}, 1, 2, 3);
// 6
```

This is an issue, since if I use something like clone() or merge() on an object, it can be overridden by the given source...

I think I should do it with the latter one.

**Comment by inf3rno** [comment]
```js
var merge = function (subject, source) {
    if (!(subject instanceof Object))
        throw new InvalidArguments();
    if (subject.merge instanceof Function)
        return subject.merge.apply(subject, toArray(arguments).slice(1));
    return shallowCopy.apply(null, arguments);
};

var shallowCopy = function (subject, source) {
    if (!(subject instanceof Object))
        throw new InvalidArguments();
    var sources = toArray(arguments).slice(1);
    for (var index in sources) {
        source = sources[index];
        if (source === undefined || source === null)
            continue;
        if (!(source instanceof Object))
            throw new InvalidArguments();
        for (var property in source)
            subject[property] = source[property];
    }
    return subject;
};
```

This is the current solution. The merge calls subject.merge or shallowCopy, so I can call shallowCopy from subject.merge() - if it is given - without recursion.

So shallowClone and deepClone should work the same way. The clone should be the only one, which calls subject.clone(), the other ones don't.

I don't necessary need both shallow and deep implementation if the deep without config does the same as the shallow does.

**Comment by inf3rno** [comment]
```js
var x = function (subject, source){
    if (subject.x)
        return subject.x.apply(subject, toArray(arguments).slice(1));
    return deepX.apply(this, arguments);
};
var deepX = function (){}; //calls shallowX
var shallowX = function (subject, source){...}; // should work only with a single one?
```
let's check the concept on real examples with nested objects:

```js
merge({x:{a:1}}, {x:{b:2}}); // need {x:{a:1,b:2}} as result
deepMerge({x:{a:1}}, {x:{b:2}});
x = shallowMerge({a:1}, {b:2});
o = shallowMerge({x:{a:1}}, {x:x})
retorno;
```

```js
clone({x:{a:1}}); // need {x:{a:1}} as result where x is cloned as well
deepClone({x:{a:1}});
x = shallowClone({a:1});
o = shallowClone({x:{a:1}})
o.x = x;
retorno;
```

so what do we have here?
```js
x = shallowMerge({a:1}, {b:2});
o = shallowMerge({x:{a:1}}, {x:x})
```
both of these must be configurable.

this is what we need:
```js
function deepMerge(o1, o2){
    return a(o1, o2);
}

function a (o1, o2){
    for (var p in o2)
        if (p == "x")
            o1[p] = b(o1[p], o2[p]);
        else
            o1[p] = o2[p];
    return o1;
}

function b(o1, o2){
    for (var p in o2)
        o1[p] = o2[p];
    return o1;
}
```

what about multiple params?
what if we use y by a or b?

```js
function b(o1, o2, o3){
    var oxs = toArray(arguments);
    oxs.shift();
    for (var index in oxs) 
        o1 = o1.y(oxs[index]);
    return o1;
}

function y(o1, ox){
    for (var property in ox)
         o1[property] = ox[property];
    return o1;
}
```
or
```js
function b(o1, o2, o3){
    return o1.y.apply(o1, toArray(arguments).slice(1));
}

function y(o2, o3){
    var o1 = this;
    var oxs = toArray(arguments);
    for (var index in oxs) {
        var ox = oxs[index];
        for (var property in ox)
             o1[property] = ox[property];
    }
    return o1;
}
```

what is the difference?
1.) by the first one the y() - which is a custom method - is much simpler, but if it overrides the original y() on the fly, then the next ox will be processed with the new y(). there is no general way to prevent this.

```js
o1 = {merge: function (){}};
o2 = {merge: function (){}};
o3 = {merge: function (){}};
o4 = {merge: function (){}};
merge(o1, o2, o3, o4);
// o1.merge(o2)
// o2.merge.call(o1, o3)
// o3.merge.call(o1, o4)
```

it is very hard to debug if an error occurs somewhere. what we will know, that somewhere in the loop we have a merge which fails. that's all. this might be a small exaggeration because the file and line will be provided if it throws an error. if not, then we face a real debug nightmare.

2.) by the second one the y() is more complicated, but it is less error prone: if it overrides the original y() on the fly, then it won't affect anything on the fly.

```js
o1 = {merge: function (){}};
o2 = {merge: function (){}};
o3 = {merge: function (){}};
o4 = {merge: function (){}};
merge(o1, o2, o3, o4);
// o1.merge(o2)
// o1.merge(o3)
// o1.merge(o4)
```

except if we have other type of recursion in the process e.g. 

```js
y(o1.x, o2.x, o3.x);
y(o1.z, o2.z, o3.z);
    y(o1.z.x, o2.z.x, o3.z.x); // where o{i}.z.x is a reference to o{i}.x
```

this is a very rare case...

I think I should use the second solution with a helper that helps to convert the first one to it:

```js
o1 = {merge: multiply(function (ox){}};
o2 = {merge: function (){}};
o3 = {merge: function (){}};
o4 = {merge: function (){}};
merge(o1, o2, o3, o4);
// o1.merge(o2)
// o1.merge(o3)
// o1.merge(o4)
```

the problem that this is error prone on a different level... we can make the same mistakes as 1.) by multiplying functions which we should not.

**Comment by inf3rno** [comment]
let's check jquery extend:

```js
jQuery.extend = jQuery.fn.extend = function () {
    var options, name, src, copy, copyIsArray, clone,
        target = arguments[0] || {},
        i = 1,
        length = arguments.length,
        deep = false;
// Handle a deep copy situation
    if (typeof target === "boolean") {
        deep = target;
// Skip the boolean and the target
        target = arguments[i] || {};
        i++;
    }
// Handle case when target is a string or something (possible in deep copy)
    if (typeof target !== "object" && !jQuery.isFunction(target)) {
        target = {};
    }
// Extend jQuery itself if only one argument is passed
    if (i === length) {
        target = this;
        i--;
    }
    for (; i < length; i++) {
// Only deal with non-null/undefined values
        if ((options = arguments[i]) != null) {
// Extend the base object
            for (name in options) {
                src = target[name];
                copy = options[name];
// Prevent never-ending loop
                if (target === copy) {
                    continue;
                }
// Recurse if we're merging plain objects or arrays
                if (deep && copy && ( jQuery.isPlainObject(copy) ||
                    (copyIsArray = jQuery.isArray(copy)) )) {
                    if (copyIsArray) {
                        copyIsArray = false;
                        clone = src && jQuery.isArray(src) ? src : [];
                    } else {
                        clone = src && jQuery.isPlainObject(src) ? src : {};
                    }
// Never move original objects, clone them
                    target[name] = jQuery.extend(deep, clone, copy);
// Don't bring in undefined values
                } else if (copy !== undefined) {
                    target[name] = copy;
                }
            }
        }
    }
// Return the modified object
    return target;
};
```

this is a simple deepCopy, does not help :S

**Comment by inf3rno** [comment]
The current clone:

```js
var clone = function (subject) {
    if (typeof (subject) != "object" || subject === null)
        return subject;
    if (subject instanceof Array)
        return subject.slice();
    if (subject instanceof Date)
        return new Date(subject);
    if (subject instanceof RegExp)
        return new RegExp(subject);
    if (subject.clone instanceof Function)
        return subject.clone();
    return Object.create(subject);
};
```

the simplest deepClone would be:

```js
    if (subject instanceof Array){
        var result = subject.slice();
        for (var index in result)
            result[index] = clone(result[index]);
    }
```
    
```js   
        var result = Object.create(subject);;
        for (var index in result)
            result[index] = clone(result[index]); 
```

why not to implement it in a single function?

I mean I can add a config object to override self calling by merge...

```js
var merge = function (subject, source) {
    if (!(subject instanceof Object))
        throw new InvalidArguments();
    if (subject.merge instanceof Function)
        return subject.merge.apply(subject, toArray(arguments).slice(1));
    return shallowCopy.apply(null, arguments);
};
```

```js
    merge: function (source) {
        var parameters = toArray(arguments);
        parameters.unshift(this);
        return shallowCopy.apply(null, parameters);
    },
```

I could use an override flag:

```js
merge.apply({"":merge}, parameters);
```
this is a bit ad-hoc solution, but with this it will be evident to the engine, that it has to use the shallowCopy instead of a recursive obj.merge() call.

**Comment by inf3rno** [comment]
what about calling the same merge on the attributes?

```js
merge.apply({
    "": function (subject, source){}, // which I call on the root
    "*": function (property, sourceProperty){}, // which I call on the single level deep properties
    "**": function (nestedProperty, nestedSourceProperty){} // which I call on the nested properties
});
```
how to transform this context into a single or multi level deep context?

single level:
```js
{
    "": function (property, sourceProperty){}, // which I call on the single level deep properties
    "**": function (nestedProperty, nestedSourceProperty){} // which I call on the nested properties
}
```

multi level:
```js
{
    "**": function (nestedProperty, nestedSourceProperty){} // which I call on the nested properties
}
```

wouldn't it be better to use rules?

```js
[
    {
        test: function (){
            return true;
        },
        exec: function (subject, source){
            //...
        },
        inherit: 0
    },
    // ....
]
```

it would be more flexible, but harder to configure.

**Comment by inf3rno** [comment]
I think this is the same concept I always use, it is better to write code than writing config files/objects.

**Comment by inf3rno** [comment]
Let's check a real example with this:

```js
    build: function () {
        Base.prototype.build.call(this);
        this.frames = clone(this.frames);
    },
    merge: function (source) {
        var sources = [];
        for (var sourceIndex in arguments) {
            source = arguments[sourceIndex];
            if (source === undefined || source === null)
                continue;
            if (!(source instanceof Object))
                throw new InvalidArguments();

            if (source.frames !== undefined) {
                if (!(source.frames instanceof Array))
                    throw new StackTrace.StackFramesRequired();
                for (var frameIndex in source.frames)
                    if (!(source.frames[frameIndex] instanceof StackFrame))
                        throw new StackTrace.StackFrameRequired();
                this.frames.push.apply(this.frames, source.frames);
            }
            var backup = {
                frames: this.frames
            };
            shallowCopy(this, source);
            shallowCopy(this, backup);
        }
        return this;
    },
```

**Comment by inf3rno** [comment]
```js
    build: function () {
        Base.prototype.build.call(this);
        deepClone.call(function (context, subject){
            if (this.root(context))
                this.next(context, subject);
            if (this.path(context, "frames"))
                this.next(context, clone(subject));
            else
                this.done(context, subject);
        }, this)
    },
```

hmm this is more about traversing.

isn't it possible to use a cascade solution here? from matcher to matcher?

```js
    build: function () {
        Base.prototype.build.call(this);
        cascade.call(
            this, 
            function (done, next, subject){
                next(subject.frames);
            }, 
            function (done, next, frames){
                done(Object.create(frames));
            }
        );
    },
```
this works too, but this is a too easy example to use complicated terms here...
 let's check the merge with cascade:

```js
    merge: function (source) {
        carcade.call([this, source1, source2, ...], 
            function (cascade, root){
                cascade.next(root.frames)
            },
            function (cascade, frames){
                if (!(frames instanceof Array))
                    throw new StackTrace.StackFramesRequired();
                cascade.each(frames);
            },
            function (cascade, frame){
                if (!(frame instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
                cascade.push(frame);
            },
            function (cascade, root, frames){
                cascade.merge(root, {frames: frames});
            }
        );
    },
```

not bad. it traverses the tree of the subject and sources in parallel, so I'll need to rewrite every single step hmmm btw it is possible to customize:

```js
    merge: function (source) {
        carcade.call([this, source1, source2, ...], 
            function (cascade, root){
                cascade.next(root.frames, function (subject, source1, source2){
                    var sources = toArray(arguments).slice(1);
                    for (var index in sources) {
                        var source = sources[index];
                        blahhhhhhhhhhhhh :D
                    }
                });
            },
            function (cascade, frames){
                if (!(frames instanceof Array))
                    throw new StackTrace.StackFramesRequired();
                cascade.each(frames);
            },
            function (cascade, frame){
                if (!(frame instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
                cascade.push(frame);
            },
            function (cascade, root, frames){
                cascade.merge(root, {frames: frames});
            }
        );
    },
```

I think it is too complicated...

what I really need here:

 - traverse multiple trees parallel
 - save the results in an array
 - reuse the results when needed

```js
merge: function (source) {
     var sourceFrames = traverse.call(this, toArray(arguments), function (source){
            if (source === undefined || source === null)
                return ;
            if (!(source instanceof Object))
                throw new InvalidArguments();
            if (source.frames === undefined) 
                return;
            if (!(source.frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
            for (var index in source.frames)
               if (!(source.frames[index] instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
            return source.frames;
     });
     traverse.call(this, sourceFrames, function (sourceFrames){
          this.frames.push.apply(this.frames, frames);
     });
     var frames = this.frames;
     shallowCopy(this, source);
     this.frames = frames ;
}
```
this is simple map + each, so I reinvented the wheel :D btw this does not solve anything I got the same code

what is custom by this code?

```js
            if (source === undefined || source === null)
                return ;
            if (!(source instanceof Object))
                throw new InvalidArguments();
```

```js
            if (source.frames === undefined) 
                return;
            if (!(source.frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
```

```js
               if (!(source.frames[index] instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
```

```js
this.frames.push.apply(this.frames, source.frames);
shallowCopy(this, source, {frames: this.frames});
```

**Comment by inf3rno** [comment]
The deepCopy part is

```js
this.frames.push.apply(this.frames, source.frames);
shallowCopy(this, source, {frames: this.frames});
```

the others have only validation purposes.

what I really want is something like this:

```js
shallowCopy(this, source, {frames: yyy(this.frames, source.frames)});
```
or more like:

```js
xxx({frames: yyy(this.frames, source.frames)}, this, source);
```

the traversing part would be important, because I would be able to do something like this:


```js
xxx({frames: yyy(function (o){return o.frames;})}, this, source);
```
hmmm

```js
deepCopy({
    frames: [
        function (frames, frame){
               if (!(frame instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
               frames.push(frame);
        }
    ]
}, this, source);
```

a little bit ad-hoc by iterating the array, but that's okay.

**Comment by inf3rno** [comment]
```js
deepCopy({
    frames: [
        function (frames, frame){
               if (!(frame instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
               frames.push(frame);
        }
    ]
}, {
    frames: function (frames){
        if (!(frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
        // I can throw here by undefined if I want...
        return frames;
    }
}, this, source);
```

what if I want check both the items and the array?

```js
deepCopy({
    frames: [
        function (frames, frame){
               frames.push(frame);
        }
    ]
}, {
    frames: [
        function (frames){
            if (!(frames instanceof Array))
                    throw new StackTrace.StackFramesRequired();
            // I can throw here by undefined if I want...
            return frames;
        },
        function (frame){
               if (!(frame instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
        }
    ]
}, this, source);
```
wouldn't it be simpler with way?

```js
deepCopy({
    frames: [
        {
            items: [
                function (frames, frame) {
                    frames.push(frame);
                },
                function (frame) {
                    if (!(frame instanceof StackFrame))
                        throw new StackTrace.StackFrameRequired();
                }
            ]

        },
        function (frames) {
            if (!(frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
        }
    ]
}, this, source);
```


```js
deepCopy({
    frames: {
        type: Array,
        items: [
            function (frames, frame) {
                frames.push(frame);
            },
            function (frame) {
                if (!(frame instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
            }
        ],
        check: function (frames) {
            if (!(frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
        }
    }
}, this, source);
```

```js
deepCopy({
    frames: {
        type: Array,
        items: {
            item: function (frames, frame) {
                frames.push(frame);
            },
            check: function (frame) {
                if (!(frame instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
            }
        },
        check: function (frames) {
            if (!(frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
        }
    }
}, this, source);
```

```js
deepCopy({
    frames: [
        [
            function (frames, frame) {
                frames.push(frame);
            },
            function (frame) {
                if (!(frame instanceof StackFrame))
                    throw new StackTrace.StackFrameRequired();
            }
        ],
        function (frames) {
            if (!(frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
        }
    ]
}, this, source);
```
this should be nested Array instead

```js
deepCopy({
    frames: [
        function (frames, frame) {
            if (!(frame instanceof StackFrame))
                throw new StackTrace.StackFrameRequired();
            frames.push(frame);
        },
        function (frames) {
            if (!(frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
        }
    ]
}, this, source);
```

this is okay.

**Comment by inf3rno** [comment]
```js
deepCopy({
    frames: [
        function (frames, frame) {
            if (!(frame instanceof StackFrame))
                throw new StackTrace.StackFrameRequired();
            frames.push(frame);
        },
        function (frames) {
            if (!(frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
        }
    ]
}, this, source);
```
what's the problem here?

1.) I don't have a way to keep to override the original frames. I don't think this is a real problem. if I use shallowCopy it is evident that I want to keep the original one. So by deepCopy-ing frames, I should keep the original frames object. If I don't want to keep it, I have to add a deepClone to the build.

2.) What if I just want to check whether the frames is an Array?

```js
deepCopy({
    frames: function (frames) {
        if (!(frames instanceof Array))
            throw new StackTrace.StackFramesRequired();
    }
}, this, source);
```

if it returns undefined, then it will be simply deepCopied with push.

What if I want to use nested Array?

```js
deepCopy({
    frames: [
        [
            function (frame, x) {
                frame.push(x);
            },
            function (frames, frame) {
                if (!(frame instanceof Array))
                    throw new StackTrace.StackFrameRequired();
            }
        ],
        function (frames) {
            if (!(frames instanceof Array))
                throw new StackTrace.StackFramesRequired();
        }
    ]
}, this, source);
```

what if one of them returns something?
good question, I don't know... it would probably override the basic deepCopy logic somehow, but this is not clear yet... let's check another example.

**Comment by inf3rno** [comment]
Subscription

```js
    merge: function (source) {
        for (var sourceIndex in arguments) {
            source = arguments[sourceIndex];
            if (source === undefined || source === null)
                continue;
            if (!(source instanceof Object))
                throw new InvalidArguments();

            if (source.items !== undefined) {
                if (!(source.items instanceof Array))
                    throw new Subscription.ItemsRequired();
                if (source.items.length)
                    this.addAll.apply(this, source.items);
            }

            var backup = {
                items: this.items
            };

            shallowCopy(this, source);
            shallowCopy(this, backup);
        }
        return this;
    },
```

**Comment by inf3rno** [comment]
```js
deepCopy({
    items: [
        function (items, item){
             this.add(item);
        }
    ]
}, this, source);
```

or

```js
deepCopy({
    items: function (subject, items){
        if (!(items instanceof Array))
            throw new Subscription.ItemsRequired();
        subject.addAll.apply(subject, items);
    }
}, this, source);
```

what about the other properties?
I don't think they should be deepCopied by default. what if I want to deepCopy everything?

```js
deepCopy(function (subject, value, property){
    deepCopy(arguments.callee, subject[property], value);
}, this, source);
```
or something like this...

what about the return values?

```js
deepCopy(function (subject, value){
    return subject + value;
}, 1,2,3); // 6 probably...
```
but I think this should not be supported, however

```js
deepCopy(
    {
        number: function (subject, value, property) {
            subject[property] += value;
        }
    },
    {
        number: 1 // this will be 6
    },
    {
        number: 2
    },
    {
        number: 3
    }
);
```

**Comment by inf3rno** [comment]
```js
deepClone({
    items: function (items){
        return shallowClone(items);
    }
});
```

or if I want to use a nested deepClone:

```js
deepClone(function (value){
    return deepClone(value);
});
```

I think the specification is more or less complete, I start to write tests, and we'll see.

**Comment by inf3rno** [comment]
what if I want to use the built-in clone of a value?

```js
deepClone({
    items: function (items){
        return items.clone();
    }
});
```

or

```js
deepClone({
    items: function (items){
        return clone(items);
    }
});
```

**Comment by inf3rno** [comment]
```js
deepCopy({
    items: function (subject, items){
        merge(subject.items, items);
    }
}, this, source);
```

**Comment by inf3rno** [comment]
ohh I haven't decided yet to use only a single or multiple sources.

**Comment by inf3rno** [comment]
I have to use multiple, it is less error prone, and shallowCopy and deepCopy will be good supporting functions of multi-source merging.

The problem:

this is very clear:
```js
deepCopy({
    items: function (subject, items){
        if (!(items instanceof Array))
            throw new Subscription.ItemsRequired();
        subject.addAll.apply(subject, items);
    }
}, this, source);
```

this is less clear:
```js
deepCopy({
    items: function (subject, items, items2, items3){
        var itemsArray = toArray(arguments).slice(1);
        for (var index in itemsArray) {
            var items = itemsArray[index];
            if (!(items instanceof Array))
                throw new Subscription.ItemsRequired();
            subject.addAll.apply(subject, items);
        }
    }
}, this, source);
```

and there is no difference in the outcome here...
isn't there a way to solve the problem with merge somehow differently?

I don't think there is a real problem by merge since I don't use it by deepCopy. I use only a shallowCopy on the properties in where I use the config to copy the values.

**Comment by inf3rno** [comment]
I tried with this to write Wrapper, but it has an infinite loop:

```js
    merge: function (source) {
        return deepCopy(this, toArray(arguments), {
            preprocessors: [
                function (preprocessors, preprocessor) {
                    if (!(preprocessor instanceof Function))
                        throw new Wrapper.PreprocessorRequired();
                    preprocessors.push(preprocessor);

                },
                function (subject, preprocessors) {
                    if (!(preprocessors instanceof Array))
                        throw new Wrapper.ArrayRequired();

                }
            ],
            done: function (subject, done) {
                if (!(done instanceof Function))
                    throw new Wrapper.FunctionRequired();
                return done;
            },
            algorithm: function (subject, algorithm) {
                if (!(algorithm instanceof Function))
                    throw new Wrapper.FunctionRequired();
                return algorithm;
            },
            properties: function (subject, properties) {
                if (!(properties instanceof Object))
                    throw new Wrapper.PropertiesRequired();
                shallowCopy(subject.properties, [properties]);
            }
        });
    },
```

on paper it looks nice, but in the practice .... :S 
the original one is:

```js
    merge: function (source) {
        for (var sourceIndex in arguments) {
            source = arguments[sourceIndex];
            if (source === undefined || source === null)
                continue;
            if (!(source instanceof Object))
                throw new InvalidArguments();

            if (source.preprocessors !== undefined) {
                if (!(source.preprocessors instanceof Array))
                    throw new Wrapper.ArrayRequired();
                for (var preprocessorIndex in source.preprocessors)
                    if (!(source.preprocessors[preprocessorIndex] instanceof Function))
                        throw new Wrapper.PreprocessorRequired();
                this.preprocessors.push.apply(this.preprocessors, source.preprocessors);
            }

            if (source.done !== undefined && !(source.done instanceof Function))
                throw new Wrapper.FunctionRequired();

            if (source.algorithm !== undefined && !(source.algorithm instanceof Function))
                throw new Wrapper.AlgorithmRequired();

            if (source.properties !== undefined) {
                if (!(source.properties instanceof Object))
                    throw new Wrapper.PropertiesRequired();
                shallowCopy(this.properties, [source.properties]);
            }
            shallowCopy(this, [source, {
                preprocessors: this.preprocessors,
                properties: this.properties
            }]);
        }
        return this;
    },
```

it does not have an infinite loop because of merge collision, so there is something in the deepCopy which causes this... I have to write more unit tests, maybe I find a bug or maybe it will be easier to debug...

**Comment by inf3rno** [comment]
the wrapper is too frequently used, so it will be hard to debug an infinite loop. it is much easier to do something similar in a test.

**Comment by inf3rno** [comment]
```js
    merge: function (source) {
        return deepCopy(this, toArray(arguments), {
            items: function (subject, items) {
                console.log(items)
                if (!(items instanceof Array))
                    throw new Subscription.ItemsRequired();
                subject.addAll.apply(subject, items);
            }
        });
    },
```

this does not work as well.. it passes an item instead of an Array of items. I have to investigate this and rethink the options parameter, because it is not obvious

**Comment by inf3rno** [comment]
in this case:

```js
            else if (options.hasOwnProperty(property))
                deepCopy(subject[property], [value], options[property]);
```

this runs... which means that 
```js 
deepCopy(subscription.items, [new Component()], function (subject, items) {
    if (!(items instanceof Array))
        throw new Subscription.ItemsRequired();
    subject.addAll.apply(subject, items);
});
```
is called.

**Comment by inf3rno** [comment]
after that this is running:

```js
                if (callback) {
                    var result = callback(subject, value, property);
                    if (result !== undefined)
                        subject[property] = result;
                }
```

which means

```js
function (subject, items) {
    if (!(items instanceof Array))
        throw new Subscription.ItemsRequired();
    subject.addAll.apply(subject, items);
}(items, new Component());
```

it's obvious that it won't work properly...

**Comment by inf3rno** [comment]
with this hax it is possible to fix more or less:
```js
    merge: function (source) {
        return deepCopy(this, toArray(arguments), {
            items: function () {
            },
            "": function (subject, source) {
                if (!source || !source.items)
                    return
                if (!(source.items instanceof Array))
                    throw new Subscription.ItemsRequired();
                subject.addAll.apply(subject, source.items);
            }
        });
    },
```

but I think I should give the options differently than now.

**Comment by inf3rno** [comment]
```js
deepCopy(a,b,{
    x: function (subject, x){
        if (x instanceof Array)
            subject.x.push.apply(subject.x, x);
    }
});
```
e.g. for pushing an Array this is how it should work...

what if I want to check the items?

```js
deepCopy(a,b,{
    x: [
        function (x, item){
            if (!(item instanceof Component))
                throw new InvalidArguments();
        },
        function (subject, x){
            if (!(x instanceof Array))
                throw new ArrayRequired();
            subject.x.push.apply(subject.x, x);
        }
    ]
});
```
the problem here that the push runs before checking the items...

let's assume that we have 2 type of item matchers

```
[...] - matches every item in the Array or Object
{0: ..., x: ...} - matches individual items in the Array or Object (usually the latter)
```

so

```js
[function (items, item, key){}]
```
is called on each item in the sources

what about

```js
function (subject, items, property){}
```
it should be called by

```js
{
    x: function (subject, x, property){}
}
```
so by individual items

now this is a big difference...

```js
[
    function (items, item, key){},
    function (subject, items, property){}
]
```
I think this should be the default by calling every items in an Array or an Object

e.g.

```js
{
    items: [
        function (items, item, key){},
        function (subject, items, property){}
    ]
}
```
I don't think this is consistent...

why don't use:

```js
{
    items: [
        function (items, item, key){},
    ],
    "": function (subject, items, property){}
}
```
because that would run on every property, not just on the items property...

so this is the same as

```js
{
    items: function (subject, items, property){}
}
```
if we don't want to check each item

if we want to check each item, we have a problem...

we should use:

```js
{
    items: [
        function (items, item, key){},
        function (subject, items, property){before()},
        function (subject, items, property){after()},
    ]
}
```

so it is possible to merge the whole array after checked...
another solution:

```js
{
    items: [
        function (items, item, key){},
        function (subject, items, property, next){
            before()
            next()
            after()
        }
    ]
}
```
this is simpler, but harder to implement... I'll maybe do this, I don't know yet.

**Comment by inf3rno** [comment]
What if we want to check the sources?

```js
var options = [
    function next (subject, item, index){},
    function before (subject, source, index){},
    function after (subject, source, index){},
];
var a = [1,2,3];
var b = [4,5,6,7];
var c = [7,8,9,10,11];
deepCopy(a, [b, c], options);
/*
before(a, b, i)
next(a, b[j], j)
// ...
after(a, b, i)

before(a, b, i)
next(a, b[j], j)
// ...
after(a, b, i)

//...
*/
```

**Comment by inf3rno** [comment]
what is the situation by individual properties, how to add before and after to them?

```js
{
    "": [
        function next (subject, item, index){},
        function before (subject, source, index){},
        function after (subject, source, index){},
    ],
    x: options
}

/*
before(a, b, i)
next(a, b[j], j)
// ...
after(a, b, i)

before(a, b, i)
next(a, b[j], j)
// ...
after(a, b, i)

//...
before(a, b, i)
deepCopy(options, a.x, [b.x])
after(a, b, i)
*/
```

I think empty string is the most rarely used property, so it is good enough. another option is "*" or @before @after, etc...

```js
{
    "@": {
        next: function (subject, item, index){},
        before: function (subject, source, index){},
        after: function (subject, source, index){},
    },
    x: options
}
```

```js
[
    nextOptions,
    {
        before: function (subject, source, index){},
        after: function (subject, source, index){},
    }
]
```

```js
[
    nextOptions,
    before,
    after
]
```

**Comment by inf3rno** [comment]
what about:

```js
{
    "@collection": {
        before: function (subject, source, index){},
        after: function (subject, source, index){},
    },
    "@item": function (subject, item, index){},
    x: function (){}
}
```
I think this is nice...

```js
{
    "@source": {
        "@before": function (subject, source, index){},
        "@after": function (subject, source, index){},
    },
    "@property": function (subject, value, property){},
    x: function (){}
}
```

or simply:

```js
{
    "@source": function (subject, source, index){},
    "@property": function (subject, value, property){},
    x: function (){}
}
```

if before and after not needed, just before

```js
{
    "@source:before": function (subject, source, index){},
    "@source:after": function (subject, source, index){},
    "@property": function (subject, value, property){},
    x: function (){}
}
```

**Comment by inf3rno** [comment]
```js
{
    "@": {
        source: {
            before: function (subject, source, index){},
            after: function (subject, source, index){},
        },
        property: : function (subject, value, property){}
    },
    x: function (){}
}
```

**Comment by inf3rno** [comment]
```js
[[[f]]]
```
```js
{"@property": {"@property": {"@property": f}}}
```

**Comment by inf3rno** [comment]
```js
function (){}
```
can mean both

```js
{
    "@source": function (){}
}
```

and 

```js
{
    "@property": function (){}
}
```

**Comment by inf3rno** [comment]
by root the source is not too logical, but by deeper objects it can be...

```js
{
    x: function (subject, x){subject.x = x}
}
```
so I think the

```js
function (){}
```
should mean @source

and the 
```js
[function (){}]
```
should mean @property

**Comment by inf3rno** [comment]
```js
if (options instanceof Function)
    options = {"@source": options};
else if (options instanceof Array)
    options = {"@property": options[0], "@source": options[1]};
var sourceHandler = options."@source" || options."@".source,
    propertyHandler = options."@property" || options."@".property;

for (var i in sources) {
    var source = sources[i];
    sourceHandler(subject, source, i);
    for (var j in source)
        propertyHandler(subject, source[j], j);
}
```

**Comment by inf3rno** [comment]
what about

```js
{
    x: function (){}
}
```
that function is a property handler

```js
propertyHandler(subject, source.x, "x");
```

if we want to use a source handler with it, then

```js
{
    x: function (){},
    "@source": function (){}
}
```
is the solution...

**Comment by inf3rno** [comment]
what if I give a collection to the property handler?

```js
{
    x: [function (){}]
}
```

this means that the x stores a collection

```js
deepClone(subject.x, [source.x], function (subjectX, sourceX, sourceIndex){});
```

```js
{
    x: [function (){}, function (){}]
}
```

```js
deepClone(subject.x, [source.x], [
 function (subjectX, sourceX, sourceIndex){}
 function (subjectX, sourceX[property], property){}
]);
```
which is the same as:

```js
deepClone(subject.x, [source.x], {"@property": function (){}, "@source": function (){}});
```
or in general

```js
{
    x: [function (){}, function (){}]
}
```

I think I have talked about every aspect of the problem

**Comment by inf3rno** [comment]
is it possible to use callback by the `@source`?

```js
for (var i in sources) {
    var source = sources[i];
    var callback = function (){
        for (var j in source)
            propertyHandler(subject, source[j], j);
    };
    sourceHandler(subject, source, i, callback);
}
```
it is possible

```js
var isCalled = false;
var callback = function (){
    isCalled = true;
    for (var j in source)
        propertyHandler(subject, source[j], j);
};
for (var i in sources) {
    var source = sources[i];
    sourceHandler(subject, source, i, callback);
    if (!isCalled && propertyHandler)
        callback();
}
```

that's all...

I'll implement it this way.

**Comment by inf3rno** [comment]
These are the json-ld reserved properties:

```js
@context
@id
@value
@language
@type
@container
@list
@set
@reverse
@index
@base
@vocab
@graph
```

so the @source and the @property is okay

**Comment by inf3rno** [comment]
```js
deepClone({"@property": 1, "@source":2}, [{@property: 3}], {"@property": {default: function (){blabla}})
```
I don't think we need this, I mean if somebody wants to use deepClone with these properties, then they will realize very soon, that they can't do that because I use these properties for different purposes. So this is not so error prone, and it will be most probably a rare case. What I have to do is checking the property name by this:

```js
            else if (options.hasOwnProperty(property))
                deepCopy(subject[property], [value], options[property]);
```

so it won't cause any issue if the sources contain those properties.

**Comment by inf3rno** [comment]
stacktrace merging:

```js
    frames: [],
    string: undefined,
    build: function () {
        Base.prototype.build.call(this);
        this.frames = clone(this.frames);
    },
    merge: function (source) {
        return deepCopy(this, toArray(arguments), {
            frames: [
                function (frames, frame, index) {
                    if (!(frame instanceof StackFrame))
                        throw new StackTrace.StackFrameRequired();
                    frames.push(frame);
                },
                function (subject, frames, property) {
                    if (!(frames instanceof Array))
                        throw new StackTrace.StackFramesRequired();
                }
            ]
        });
    },
```

the config:

```js
        {
            frames: [
                function (frames, frame, index) {
                    if (!(frame instanceof StackFrame))
                        throw new StackTrace.StackFrameRequired();
                    frames.push(frame);
                },
                function (subject, frames, property) {
                    if (!(frames instanceof Array))
                        throw new StackTrace.StackFramesRequired();
                }
            ]
        }
```

with the current specification:

what do we want there?
 - checking the frames source (before)
 - checking the items (next)
 - copying the items after the checks are done (after)

```js
{
	frames: {
		"@source": function (subject, frames, property, each) {
			if (!(frames instanceof Array))
				throw new StackTrace.StackFramesRequired();
			each();
			subject.frames.push.apply(subject.frames, frames);
		},
		"@property": function (frames, frame, index) {
			if (!(frame instanceof StackFrame))
				throw new StackTrace.StackFrameRequired();
		}
	}
}
```

```js
{
	frames: {
		"@source": function (subject, frames, property, each) {
			if (!(frames instanceof Array))
				throw new StackTrace.StackFramesRequired();
			var filteredFrames = each(function (frame){if (frame ...) return frame});
			subject.frames.push.apply(subject.frames, frames);
		},
		"@property": function (frames, frame, index) {
			if (!(frame instanceof StackFrame))
				throw new StackTrace.StackFrameRequired();
		}
	}
}
```

interesting... I won't support this yet, but it is a nice idea if I implement the each/next function.

**Comment by inf3rno** [comment]
hmm maybe the `@source` and the `@property` are not the best words
 - source, property
 - once, each
 - branch, leaf
 - parent, child
etc...

I think the once and each are the best terms.

**Comment by inf3rno** [comment]
so the 

```js
{frames:[f1, f2]}
```
is equivalent to the

```js
{frames:{@once: f2, @each:f1}}
```
hmm maybe I should reverse the order.

```js
{frames:{@once: f1, @each:f2}}
```
what if only a single function given?

```js
{frames: f} // @once
{frames: [f]} // @each
{frames: [f, f2]} //?
```

I think the `@once` before `@each` is the better order. if we start to write the code which uses both, it is logical to start with the `@once` and write the `@each` only after that.

**Comment by inf3rno** [comment]
```js
f // @each
{@once: f, @each: f2}
```
in this context
 - `@once` - (subject, source, index, each)
 - `@each` - (subject, source[property], property)

so by `@each` it is probably possible to give further options

```js
{
   @once: function (subject, source, index, each){},
   @each: {
        @once: function (subject, source[property], property, each){},
        @each: function (subject[property], source[property][key], key){}
    },
    x: {
        @once: function (subject, source.x, "x", each){},
        @each: function (subject.x, source.x[key], key){}
    },
    y: function (subject, source.y, "y", each){}
}
```
there should be a default `@once` I think. so if no once given, then it is called. I think calling the `each` function must be mandatory from `@once` should the `@each` have a `next()` callback as well?

what about each parameters?

```js
each(subject.x, source.x);
next(subject.x, source.x[key], key);
```

we could add a `next()` to the `@each` parameters, but that does not really makes sense in that context I think.

what if we don't call the each()?
since asnyc is not supported, this can be detected. I think by once it is okay to call the each automatically after. what if no each and other properties given?

```js
{
   @once: function (subject, source, index, each){},
}
```
this would be the exact same as 

```js
function (subject, source, index, each){}
```

so we can infer that the developer requires a deep copy of the object so it will be the same as

```js
{
   @once: function (subject, source, index, each){},
   @each: function (subject, source[property], property){return source[property];}
}
```
or

```js
{
   @once: function (subject, source, index, each){},
   @each: function (subject, source[property], property){subject[property] = source[property];}
}
```
if you not prefer returning values...
ofc this will be solved with shallowCopy

so if a `@once` is given in an object, that means that the `@each` must be called. the only exception when the `@once` throws and exception.

what if I don't want to copy the properties in some of the cases, but I want to copy the properties in the others?

not calling the `each()` would be a solution in this cases. but it is error prone if somebody forgets to call it.

we can do a hax:

```js
each.called = true;
```

```js
each.skip();
```

but I think I won't need this. ofc. it can be implemented if this feature is needed.
I think the each should be called automatically if it is not called manually...

another possible solution:

```js
{
    @once: function (subject, source, index, each){
        each(function (subject, value, property){}); // or
        each({@once: function (subject, value, property){}, ...}); // or
        each([function (subject, value, property){}, ...]);
    }
}
```

what if a `@each` already given? then the parameter overrides it. I think it is better to throw an exception in that case, since it is inconclusive.

**Comment by inf3rno** [comment]
How to handle nested properties?

the normal behavior is

```
@once(subject, source)
@each(subject, value, property)
```

is it possible to build the nested properties the same way?

an example:
```js
{items: [0,1,2]}
[{items: [3,4,5]}, {items: [6,7,8]}]
```
I want to check whether every item is a Number and copy them at once
normally this would be:

```js
options = {
    @once: function (subject, source, index, each){
        if (!(source instanceof Array))
            throw new InvalidArguments();
        each();
        subject.push.apply(subject, source);
    },
    @each: function (subject, value, property){
        if (!isNaN(value))
            throw new InvalidArguments();
    }
}
```

hmm by nested we could use
```js
{
    items: options
}
```

in this case we have to call deepCopy by items, so it will be recursive
in the case of a Function, the items would be called with (subject, items, "items") and the result would be the new value

is it possible to do something similar by nested?

**Comment by inf3rno** [comment]
hmm a real nested copy would be

```js
{
    @each: function (subject, value, property){
        deepCopy(subject[property], value, arguments.callee);
    }
}
```

note that in this case we would change only the leaves and not the branches... to do that you have to use something like deepClone before deepCopy


```js
options = {
    @each: function (subject, value, property){
        // by identical structures or check whether the subject contains the property
        var newValue = deepClone(subject[property], ...); // not specified yet
        deepCopy(subject[property], value, options);
    }
}
```

**Comment by inf3rno** [comment]
by the `items` a similar solution would be

```js
{
    items: function (subject, items){
        deepCopy(subject.items, [items], options);
    }
}
```

so the 

```js
{items: options}
```

should be equal with this.

the return value if the callback is lost, since the deepCopy always returns the subject.items. the only way to override the subject.items, if I do it in `@once` before or after the items are processed. the `@once` cannot have return value, since it is called by every source. if we want to change the return value of the deepCopy, then I think the easiest way to add another callback

```js
{
    @return: function (subject, sources, options, once){
        subject = deepClone(subject); //...
        once(subject);
        return subject;
    }
}
```
now this example is nonsense because I can do the exact same by cloning the subject before calling the deepClone.

**by the properties I suggest the same, deepClone should be used before calling deepCopy.
so the deepCopy changes the leaves, while the deepClone clones the branches.
I think I should use a tree in the documentation to desribe the behavior of these functions.** #24

**Comment by inf3rno** [comment]
I think this is okay the way it is, I'll implement nested config objects.

**Comment by inf3rno** [comment]
fixing the previous examples
```js
{
    @each: function (subject, value, property){
        if (subject[property] instanceof Object)
            deepCopy(subject[property], value, arguments.callee);
        else
            return value;
    }
}
```

**Comment by inf3rno** [comment]
still somehow fails by testing on real life examples...
I don't understand why, it is well tested now, and it is consistent...

```js
Failures:
1) core CompositeError prototype stack returns the joined stack of the contained Errors
  Message:
    InvalidConfiguration: An array of frames is required.
  Stack:
    InvalidConfiguration An array of frames is required.
    	at Object.deepCopy.@once (d:\creation\software developer\projects\dataflower\core.js:348:27)
    	at deepCopy (d:\creation\software developer\projects\dataflower\core.js:207:26)
    	at Base.extend.merge (d:\creation\software developer\projects\dataflower\core.js:345:16)
    	at extend.init (d:\creation\software developer\projects\dataflower\core.js:230:20)
    	at new Descendant (d:\creation\software developer\projects\dataflower\core.js:85:23)
    	at createError (d:\creation\software developer\projects\dataflower\spec\core\CompositeError.spec.js:20:45)
    	at Object.<anonymous> (d:\creation\software developer\projects\dataflower\spec\core\CompositeError.spec.js:28:29)

2) core CompositeError prototype stack supports native errors as well
  Message:
    InvalidConfiguration: An array of frames is required.
  Stack:
    InvalidConfiguration An array of frames is required.
    	at Object.deepCopy.@once (d:\creation\software developer\projects\dataflower\core.js:348:27)
    	at deepCopy (d:\creation\software developer\projects\dataflower\core.js:207:26)
    	at Base.extend.merge (d:\creation\software developer\projects\dataflower\core.js:345:16)
    	at extend.init (d:\creation\software developer\projects\dataflower\core.js:230:20)
    	at new Descendant (d:\creation\software developer\projects\dataflower\core.js:85:23)
    	at Object.<anonymous> (d:\creation\software developer\projects\dataflower\spec\core\CompositeError.spec.js:94:41)
```

maybe it is too good and does not accept mock :D

**Comment by inf3rno** [comment]
Hmm another interesting thing, that the StackTrace.toString() has error and not the merge. I don't understand :S

**Comment by inf3rno** [comment]
The problem is with the 

```js
new StackTrace({
                                    toString: function () {
                                        return options.stack;
                                    }
                                })
```

which I use to mock out the real stack... the toString() is not copied or overridden with something.

**Comment by inf3rno** [comment]
it is `Object.prototype.toString.call(undefined)` so somehow calls the original toString on undefined. this can happen only if the toString is in the options. and yes it is in there. I check it with `options[property] === undefined` which does not work.

**Comment by inf3rno** [comment]
```js
propertyOptions = options[property] || options[each];
```

the problem is here
I have to check with options.hasOwnProperty(property)...

**Comment by inf3rno** [comment]
I think it is finished. I overrode the merge()-s in everywhere, it works perfectly. one thing I just realized that it is good for validation as well. so most of the configure() logic will be useless after I implemented it. another thing, that I can store the options in a property, so it can be cloned with deepClone and overridden with deepCopy. currently overriding properties like `@once` and `@each` is just partially supported. a full support should be part of the next milestone if it is needed.

**Comment by inf3rno** [comment]
I can start to work on the deepClone. It will have a similar approach. `@once` will return, if not given, then it can use Object.create() or shalllowClone, or something like that automatically.

**Comment by inf3rno** [comment]
this is the current clone:

```js
var clone = function (subject) {
    if (typeof (subject) != "object" || subject === null)
        return subject;
    if (subject instanceof Array)
        return subject.slice();
    if (subject instanceof Date)
        return new Date(subject);
    if (subject instanceof RegExp)
        return new RegExp(subject);
    if (subject.clone instanceof Function)
        return subject.clone();
    return Object.create(subject);
};
```

it should be split into shallowClone and clone just by merge. so subject.clone() would be able to use shallowClone() and deepClone() (after it is implemented)

I might want to rethink this. so far deepCopy works perfectly, but I never tried to deepCopy properties which have their own merge... I would use an options something like this with the current impl.

```js
{
    niceProperty: function (subject, value, property){
        merge(subject[property], value);
    }
}
```

**Comment by inf3rno** [comment]
Clone is used in extend

```js
    Descendant.prototype = clone(Ancestor.prototype);
```

and in Task.receive: (just for cloning an Array)

```js
        parameters = clone(parameters);
```

let's check the build codes:

Base (current)
```js
    init: function () {
        this.build();
        this.merge.apply(this, arguments);
        this.configure();
    },
    clone: function () {
        var instance = Object.create(this);
        instance.build();
        return instance;
    },
    build: function () {
        Object.defineProperty(this, "id", {
            configurable: false,
            enumerable: false,
            writable: false,
            value: id()
        });
    },
    merge: function (source) {
        return shallowCopy(this, toArray(arguments));
    },
```

StackTrace
```js
    frames: [],
    build: function () {
        Base.prototype.build.call(this);
        this.frames = clone(this.frames);
    },
```

Wrapper
```js
    preprocessors: [],
    done: function () {
        return toArray(arguments);
    },
    algorithm: function (wrapper) {
        return function () {
            return wrapper.done.apply(this, arguments);
        };
    },
    properties: {},
    build: function () {
        Base.prototype.build.call(this);
        this.preprocessors = clone(this.preprocessors);
        this.properties = clone(this.properties);
    },
```

HashSet
```js
    items: {},
    init: function () {
        this.build();
        this.configure.apply(this, arguments);
    },
    build: function () {
        Base.prototype.build.call(this);
        var inheritedItems = this.toArray();
        this.items = {};
        this.addAll.apply(this, inheritedItems);
    },
```

**Comment by inf3rno** [comment]
```js
deepClone(obj); // -> Object.create(obj)
deepClone(primitece); // -> throw Error
deepClone(function (){}); // -> throw Error (most probably), it's the task of extend I guess
deepClone(obj, {property: function (subject, value, property){}}); // use the calback to clone the property
deepClone(obj, {@once: function (subject){}}); // not necessary by the root
deepClone(obj, {@each: function (subject, value, property){}}); // clone every property
deepClone(obj, [function each(){}]); // {@each: each}
deepClone(obj, [function once(){}, function each(){}]); // {@once: once, @each: each}
deepClone(obj, {property: {@each: function each(){}, @once: function once(){}}}); // clones the property with once, or if not given with Object.create. clones the sub-properties with each
deepClone(obj, [function once1(){}, [function once2(){}, function each(){}]]); // clones the root with once1, clones @each with once2 and clones @each@each with each
```

it always uses the return values of once() if none/undefined given, then it throws Error.

**Comment by inf3rno** [comment]
I think this is enough as specification. I'll solve cloning properties like `"@once"` and `"@each"` in #72

**Comment by inf3rno** [comment]
I suggest to throw Error if somebody wants to use these reserved properties in the source objects. that will prevent issues before writing the specification about how to solve this.

**Comment by inf3rno** [comment]
Let's concentrate on Arrays and Objects. What deepClone means by them?

by Objects: Object.create(subject), by Arrays: subject.slice() at first
after that iteration over the properties/items and clone of them one by one with `@each`.

so I have to exclude RegExp, Date, etc... at first, since those are shallowClone targets.

hmm what will drive the iteration? 
by deepMerge it is driven by the sources. I think by deepCopy it should be driven by the config. 

what if a property is defined by the config but not by the subject?
it can be ignored or it can throw error. should I use a `@required`/`@optional` property to describe this?

**Comment by inf3rno** [comment]
should deepClone accept only objects as subject? no, but yes if a config given. the traversing will be driven by the config.

**Comment by inf3rno** [comment]
I think I already implemented something like this by inheritance.js in a snowy winter day maybe a year ago.

**Comment by inf3rno** [comment]
```js
deepClone({
    a: {b:1}
}, {
    a: function (subject, a, property, path){
        // automatically checks whether the property is set
        if (!(value instanceof Object) // not necessary if it is safe to assume that it is an Object.
            throw new InvalidArguments.Nested({path: path});
        return Object.create(a); // or shallowCopy(a)
    }
});
```

```js
deepClone({
    a: {b:1}
}, {
    a: {
        "@optional": true, // if not defined then it skips
        "@once": function (a, each, path){
            // automatically checks whether the property is set
            if (!(value instanceof Object) // not necessary if it is safe to assume that it is an Object.
                throw new InvalidArguments.Nested({path: path});
            var instance = Object.create(a); // or shallowCopy(a)
            each(instance);
            return instance;
        },
        b: function (instance, b, property, path){
            return b+1;
        },
        "@each": function (instance, value, property, path){
            return shallowClone(value);
        }
    }
});
```

if no `@once` given, then it calls Object.create() or array.slice() automatically and calls the each after that on the new instance. I think this is good enough...

**Comment by inf3rno** [comment]
```js
deepClone(a, {x:null})
```
since x is defined here, it will clone the x property as well with deepClone() or shallowClone().

**Comment by inf3rno** [comment]
so by options

```js
undefined // equals to 
null // equals to
{}
```

**Comment by inf3rno** [comment]
what if a single function given as options? it is the same as `{"@once": f}` I guess

**Comment by inf3rno** [comment]
I checked both the tests and the code of deepMerge and found nothing about having Function options.

by nested options:
```js
            if (property == each || property == once)
                throw new InvalidArguments.Nested({path: path});
            propertyOptions = options[each];
            if (options.hasOwnProperty(property))
                propertyOptions = options[property];
            if (propertyOptions === undefined)
                subject[property] = source[property];
            else if (propertyOptions instanceof Function) {
                result = propertyOptions(subject, source[property], property, path);
                if (result !== undefined) {
                    subject[property] = result;
                    delete (result);
                }
            }
            else
                deepMerge(subject[property], [source[property]], propertyOptions, path);
```
to clear

```js
            if (propertyOptions === undefined)
                copy-the-property();
            else if (propertyOptions instanceof Function) 
                call-propertyOptions-on-it();
            else
                deepMerge-with-propertyOptions-as-options();
```

so it does not support anything special with Function options and it does not convert it to `@once` either... that would be ambiguous since the properties use it in a different way. I should throw an Error by Function options.

**Comment by inf3rno** [comment]
hmm is it possible to fuse the deepClone and deepMerge somehow in a single function called deepCopy?

the problem here, that deepClone is config driven, while deepMerge is source driven. so they work completely differently. I could run the deepClone before the deepMerge.

what about the options? the merge options require source, while the clone options don't use the source. I will be able to use if-else if I want to distinguish between them, but I think it is better to move these things to different properties:

```js
{
    x: {
        "@clone": function (x, path){},
        "@merge": function (x, sourceValue, property){}
    }
}
```

```js
{
    x: function (subject, sourceValue, property){
        // we would clone it by every source call
    }
}
```
this won't work and would made things even harder to understand

**Comment by inf3rno** [comment]
```
{
    x: {
        "@??": function (subject, sourceValue, property){}
        "@once": function (subjectValue, path){},
        "@clone": function (subjectValue, path){},
        "@merge": function (subjectValue, sourceValue, property){}
    }
}
```
the `@once` would be probably deprecated, the `@clone` would be able to decide whether it wants to clone or not. if it gives a return value, then it clones the property. if not, then it doesn't. what is the problem here? the `@once` runs once by every source, but the `@clone` should run only a single time...

```js
{
    x: {
        "@only-a-single-time": function (subjectValue, path){}, //clone
        "@by-each-source-if-property-defined": function (subjectValue, sourceValue, index, each, path){},
        "@by-each-property-of-each-source-value": function (subjectValue, sourceSubValue, subProperty, path){}
    }
}
```
hmm this could work.

**Comment by inf3rno** [comment]
```js
{
    x: {
        "@single": function (subjectValue, path){}, //clone
        "@once": function (subjectValue, sourceValue, index, each, path){},
        "@each": function (subjectValue, sourceSubValue, subProperty, path){}
    }
}
```

```js
{
    x: {
        "@subject": function (subject, subject[property], property, merge, path){}, //clone
        "@source": function (subject[property], sources[index][property], index, items, path){},
        "@item": function (subject[property], sources[index][property][subProperty], subProperty, path){}
    }
}
```
hmm nice



```js
{
    "@subject": function (undefined, subject, undefined, merge, path){}, //clone
    "@source": function (subject, sources[index], index, items, path){},
    "@item": function (subject, sources[index][property], property, path){}
}
```
it should be rather

```js
{
    "@subject": function (subject, eachSource, path){}, //cloning subject
    "@source": function (subject, sources[index], eachProperty, path){}, //source check
    "@property": function (subject, sources[index][property], property, path){} //copying source properties
}
```

so

```js
{
    "@subject": function (subject, eachSource, path){
        //cloning subject
        eachSource(subjectClone);
        return subjectClone; 
            // optional because eachSource gets the clone in the arguments
            // necessary only if eachSource is not called
            // they can return a different value @subject can work further on the clone...
    },
    "@source": function (subject, source, eachProperty, path){
        //source check
        eachProperty();
        //bulk property merge if necessary
        return;
            // nothing to return
    },
    "@property": function (subject, value, property, path){
        //value check
        //single merge if necessary
        return newValue;
            // optional because value can be set using subject[property]
            // if both given then the return value overrides the other one
    }
}
```

this is almost like the current solution `@once` is `@source` and `@each` is `@property`.

fine, let's assume I modify the deepMerge to use this syntax.

**Comment by inf3rno** [comment]
Error
```js
123
"string"
false
```

shallow mode
```js
null
undefined
{}
[]
```

deep mode
```js
{...}
[...]
```

**Comment by inf3rno** [comment]
what will be the default behavior by shallow mode?

```
shallowClone -> Object.create
shallowMerge -> subject[property] = source[property]
```
options:
 - do nothing and return the original object
 - shallow merge if sources given
 - clone and shallow merge if sources given

it is easy to turn off cloning:

```js
{
    "@subject": echo
}
```
it is easy to turn on cloning either:

```js
{
    "@subject": shallowClone // or clone()
}
```

we have to distinguish between the properties we want to merge and the ones we want to clone and merge and the ones we want to clone but not merge

**Comment by inf3rno** [comment]
Let's check existing examples:

```js
    frames: [],
    build: function () {
        Base.prototype.build.call(this);
        this.frames = clone(this.frames);
    },
    merge: function (source) {
        return deepMerge(this, toArray(arguments), {
            frames: [
                function (frames, sourceFrames, index, each, path) {
                    if (!(sourceFrames instanceof Array))
                        throw new StackTrace.StackFramesRequired();
                    each();
                    frames.push.apply(frames, sourceFrames);
                },
                function (frames, frame, index, path) {
                    if (!(frame instanceof StackFrame))
                        throw new StackTrace.StackFrameRequired();
                }
            ]
        });
    },
```

```js
    preprocessors: [],
    done: function () {
        return toArray(arguments);
    },
    algorithm: function (wrapper) {
        return function () {
            return wrapper.done.apply(this, arguments);
        };
    },
    properties: {},
    build: function () {
        Base.prototype.build.call(this);
        this.preprocessors = clone(this.preprocessors);
        this.properties = clone(this.properties);
    },
    merge: function (source) {
        return deepMerge(this, toArray(arguments), {
            preprocessors: [
                function (preprocessors, sourcePreprocessors, index, each) {
                    if (!(sourcePreprocessors instanceof Array))
                        throw new Wrapper.ArrayRequired();
                    each();
                    preprocessors.push.apply(preprocessors, sourcePreprocessors);
                },
                function (preprocessors, preprocessor) {
                    if (!(preprocessor instanceof Function))
                        throw new Wrapper.PreprocessorRequired();
                }
            ],
            done: function (wrapper, done) {
                if (!(done instanceof Function))
                    throw new Wrapper.FunctionRequired();
                return done;
            },
            algorithm: function (wrapper, algorithm) {
                if (!(algorithm instanceof Function))
                    throw new Wrapper.AlgorithmRequired();
                return algorithm;
            },
            properties: []
        });
    },
```

```js
    items: {},
    build: function () {
        Base.prototype.build.call(this);
        var inheritedItems = this.toArray();
        this.items = {};
        this.addAll.apply(this, inheritedItems);
    },
```

```js
    merge: function (source) {
        return deepMerge(this, toArray(arguments), {
            items: function (subscription, items) {
                if (!(items instanceof Array))
                    throw new Subscription.ItemsRequired();
                subscription.addAll.apply(subscription, items);
            }
        });
    },
```

**Comment by inf3rno** [comment]
Start with the StackTrace

with `@echo`:

```js
    frames: [],
    build: function () {
        deepCopy(this, null, {
            "@subject": echo, // prevent clone
            frames: {}
        });
        Base.prototype.build.call(this);
    },
    merge: function (source) {
        return deepCopy(this, toArray(arguments), {
            "@subject": echo, // prevent cloning the trace
            frames: {
                "@subject": echo, // prevent cloning the frames
                "@source": function (frames, sourceFrames, eachProperty, path) {
                    if (!(sourceFrames instanceof Array))
                        throw new StackTrace.StackFramesRequired();
                    eachProperty();
                    frames.push.apply(frames, sourceFrames);
                },
                "@property": function (frames, frame, index, path) {
                    if (!(frame instanceof StackFrame))
                        throw new StackTrace.StackFrameRequired();
                }
            }
        });
    },
```

**Comment by inf3rno** [comment]
with `shallowClone`

```js
    frames: [],
    build: function () {
        deepCopy(this, null, {
            frames: shallowClone
        });
        Base.prototype.build.call(this);
    },
    merge: function (source) {
        return deepCopy(this, toArray(arguments), {
            frames: {
                "@source": function (frames, sourceFrames, eachProperty, path) {
                    if (!(sourceFrames instanceof Array))
                        throw new StackTrace.StackFramesRequired();
                    eachProperty();
                    frames.push.apply(frames, sourceFrames);
                },
                "@property": function (frames, frame, index, path) {
                    if (!(frame instanceof StackFrame))
                        throw new StackTrace.StackFrameRequired();
                }
            }
        });
    },
```

**Comment by inf3rno** [comment]
This last one is better, but we cannot distinguish between values we want to clone only and values we want to merge only:

```js
deepCopy({a:{b:0, c:1}}, [{a:{b:1, c:2}}], {a:{}}); // I want to merge a
deepCopy({a:{b:0, c:1}}, [{a:{b:1, c:2}}], {a:{"@subject": shallowClone}}); // I want to clone only, but not merge
```

I could define the merge logic in this case:
```js
{
    a: {
        "@subject": shallowClone, 
        "@source": function (subject, source, eachProperty){
            eachProperty.skip();
        }
    }
}
```
if it is nested there is a bigger problem...

we can add a shortcut:

```js
{
    a: {
        "@subject": shallowClone, 
        "@source": "@skip"
    }
}
```
but still do we really need to use the same config file both by cloning and merging? I don't think this is a good idea. they work on a different sub-graph, so it is like having a relative big intersection of two sets and claiming that the two sets are equal. ofc. we can glue them together with codes like this, but I don't think this is the right way to implement deepCopy.

**Comment by inf3rno** [comment]
so I stay with the original solution probably with different reserved words... the once and each are not good enough... the subject, source, property are okay.
